pkg/operation/botanist/controlplane.go

```diff
@@ -600,3 +594,79 @@ func (b *Botanist) DeleteCertBroker() error {
 	}
 	return nil
 }
+
+// CreateDeletionCheckpoint creates an empty configmap used as a checkpoint for shoot deletion. It gets created after the Kubernetes resources
+// inside a shoot have been cleaned up and the kube-apiserver is deleted.
+func (b *Botanist) CreateDeletionCheckpoint() error {
+	_, err := b.K8sSeedClient.CreateConfigMap(b.Shoot.SeedNamespace, common.DeletionCheckpoint, nil, true)
+	if apierrors.IsAlreadyExists(err) {
+		return nil
+	}
+	return err
+}
+
+// WakeUpControlPlane scales the replicas to 1 for the following deployments which are needed in case of shoot deletion:
+// * etcd-events
+// * etcd-main
+// * kube-apiserver
+// * cloud-controller-manager
+// * kube-controller-manager
+// * machine-controller-manager
+func (b *Botanist) WakeUpControlPlane() error {
+	if _, err := b.K8sSeedClient.ScaleStatefulSet(b.Shoot.SeedNamespace, fmt.Sprintf("etcd-%s", common.EtcdRoleEvents), 1); err != nil {
+		return err
+	}
+	if _, err := b.K8sSeedClient.ScaleStatefulSet(b.Shoot.SeedNamespace, fmt.Sprintf("etcd-%s", common.EtcdRoleMain), 1); err != nil {
+		return err
+	}
+	if err := b.WaitUntilEtcdReady(); err != nil {
+		return err
+	}
+
+	if _, err := b.K8sSeedClient.ScaleDeployment(b.Shoot.SeedNamespace, common.KubeAPIServerDeploymentName, 1); err != nil {
+		return err
+	}
+	if err := b.WaitUntilKubeAPIServerReady(); err != nil {
+		return err
+	}
+
+	if _, err := b.K8sSeedClient.ScaleDeployment(b.Shoot.SeedNamespace, common.KubeControllerManagerDeploymentName, 1); err != nil {
+		return err
+	}
+	if _, err := b.K8sSeedClient.ScaleDeployment(b.Shoot.SeedNamespace, common.CloudControllerManagerDeploymentName, 1); err != nil {
+		return err
+	}
+	if _, err := b.K8sSeedClient.ScaleDeployment(b.Shoot.SeedNamespace, common.MachineControllerManagerDeploymentName, 1); err != nil {
+		return err
+	}
+	return nil
+}
+
+// HibernateControlPlane hibernates the entire control plane if the shoot shall be hibernated.
+func (b *Botanist) HibernateControlPlane() error {
```

Comment:
I do not see the cloud controller manager to be scaled to 0 here.
Is this done on purpose?

Reply:
In this function only those control plane components are scaled down that cannot be immediately scaled down earlier in their respective deploy task function. CCM can, so its down-scaling happens directly there https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/controlplane.go#L442